### PR TITLE
feat(server): Garbage collector thread for project cache eviction [INGEST-1355]

### PR DESCRIPTION
**Pull request by jjbayer**
Spawn a background thread in `ProjectCache` that takes care of dropping `Project` objects, to shorten cache eviction times.

Switch from `std::collections::HashMap` to `hashbrown::HashMap` to take advantage of `drain_filter`.

### CODE REVIEW

File: relay-server/src/utils/garbage.rs [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+                // TODO: Log size of channel queue as a gauge here
+                relay_log::trace!(
+                    "Dropping object {:?} of type {}",
+                    &object as *const T,
+                    std::any::type_name::<T>()
+                );
+                drop(object);
+            }
+
+            // TODO: We create a memory leak if this thread silently crashes
```
Comment (flub) [inline relay-server/src/utils/garbage.rs]: the sender will stop working in that case.  and while the mpsc code is not the easiest to read.  I'm fairly sure that the comments indicate that the data currently in the channel is also being dropped as part of dropping the receiver. so, i **think** there is no memory leak to be concerned about

Comment (flub) [reply relay-server/src/utils/garbage.rs]: though it would result in dropping in the sending thread again

**Review by flub — APPROVED** [review]

File: relay-server/src/utils/garbage.rs (hunk 1 above)
Comment (jan-auer) [reply relay-server/src/utils/garbage.rs]: Good catch! Even when the thread panics, Rust will unwind and drop the receiver, so the Sender will start erroring. There could be corner cases where this is not possible, though we don't have to go out of our way for this.

Ideally, we would ensure that the thread then restarts. We have a similar concern for all our new services created in #1349 and #1397. Actix arbiters will restart on panic, while our new services won't. However, I think this is something that should be solved once with a generic utility, so I'm fine if we address this separately from this PR.

File: relay-server/src/utils/garbage.rs [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+        });
+
+        Self { tx }
+    }
+
+    pub fn dispose(&self, object: T) {
+        self.tx
+            .send(object)
```
Comment (jan-auer) [inline relay-server/src/utils/garbage.rs:46]: Could you run a micro-benchmark to check what the throughput with large project states is? The `Project` struct is _128_ bytes in size, so it could be more efficient to box them since the queue becomes more compact. 

Of course, the allocation of the project still exists then, and we'd have more memory fragmentation; hence a benchmark.

File: relay-server/src/utils/garbage.rs [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+    tx: mpsc::Sender<T>,
+}
+
+impl<T: Send + 'static> GarbageDisposal<T> {
+    pub fn new() -> Self {
+        let (tx, rx) = mpsc::channel();
+
+        std::thread::spawn(move || {
+            relay_log::info!("Start garbage collection thread");
+            while let Ok(object) = rx.recv() {
```
Comment (jan-auer) [inline relay-server/src/utils/garbage.rs:23]: Consider adding a cancellation token (e.g. CondVar) that stops the loop (shutting down the thread) when `GarbageDisposal` drops.

File: relay-server/src/actors/project_cache.rs [hunk 4]
```diff
@@ -77,8 +79,20 @@ impl ProjectCache {
...
-            .retain(|_, entry| entry.last_updated_at() + delta > eviction_start);
+        let config = &self.config;
+        let garbage = &self.garbage_disposal;
+        self.projects.retain(|key, entry| {
+            let keep = entry.last_updated_at() + delta > eviction_start;
+            if !keep {
+                // Useless creation of placeholder project here.
+                // This would not be necessary with nightly `drain_filter`.
+                let mut placeholder = Project::new(*key, config.clone());
+                std::mem::swap(entry, &mut placeholder);
```
Comment (jan-auer) [inline relay-server/src/actors/project_cache.rs]: Given this becomes quite unwieldy now, what if we use `hashbrown` and its [`drain_filter`](https://docs.rs/hashbrown/latest/hashbrown/struct.HashMap.html#method.drain_filter) for this?

File: relay-server/src/utils/garbage.rs (hunk 1 above)
Comment (flub) [reply relay-server/src/utils/garbage.rs]: i also kind of think this is over-thinking it.  this thread is a couple of lines of code that do not error out.  only way for that to go wrong is to panic - which clearly is a bug.  or someone starts to externally to try and corrupt this thread specifically.  that's just too far fetched.  i honestly think we can remove the todo and be done with it.

File: relay-server/src/utils/garbage.rs (hunk 3 above)
Comment (flub) [reply relay-server/src/utils/garbage.rs:23]: This thread will already stop once the sender (`GarbageDisposal`) is dropped, that's what `while let Ok(_)` does since `rx.recv()` will return `Err` if the sender is dropped.  So like, `rx` is this guard you're asking for I think?

Comment (jjbayer) [reply relay-server/src/utils/garbage.rs:23]: Dropping `GarbageDisposal` does stop the loop (by dropping `tx`), but it does not wait for the spawned thread to actually finish. I've fixed that in a more recent PR, but it does make the code less readable.

File: relay-server/src/utils/garbage.rs [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+}
+
+impl<T: Send + 'static> GarbageDisposal<T> {
+    pub fn new() -> Self {
+        let (tx, rx) = mpsc::channel();
+
+        let join_handle = std::thread::spawn(move || {
+            relay_log::debug!("Start garbage collection thread");
+            while let Ok(object) = rx.recv() {
+                // TODO: Log size of channel queue as a gauge here
```
Comment (flub) [inline relay-server/src/utils/garbage.rs]: afaik the only way to do this is by having your own `AtomicU32` to keep track of the number of entries inside it.  The other metrics option is to count on both ends, then you'll have to rates which you can subtract and no need to handle an atomic.

**Comment by flub** [comment]
I don't really understand what joining this thread buys us?  It seems a lot of complexity for no gain.

**Comment by jjbayer** [comment]
> I don't really understand what joining this thread buys us? It seems a lot of complexity for no gain.

The benefit is that I can join on it in the unit test to verify its behavior :)
But I agree about the complexity... For a long running service, it should not matter. Maybe I'll revert that change after all.

File: relay-server/src/utils/garbage.rs (hunk 2 above)
Comment (jjbayer) [reply relay-server/src/utils/garbage.rs:46]: I wrote a benchmark here: https://github.com/getsentry/relay/tree/bench/garbage-disposal (did not merge that branch because it uses some quick and dirty imports).

TL;DR boxing the `Project` does not seem to bring an advantage.

```
evict                   time:   [3.6439 us 3.6486 us 3.6536 us]                  
evict_defer             time:   [1.9363 us 1.9732 us 2.0148 us]                         
evict_convert_to_boxes  time:   [2.0021 us 2.0390 us 2.0775 us]                                    
evict_boxes             time:   [2.0992 us 2.2096 us 2.3362 us]
```

File: relay-server/src/utils/garbage.rs (hunk 3 above)
Comment (flub) [reply relay-server/src/utils/garbage.rs:23]: when do you need to wait for the thread to finish?  If this is for tests only I'd be tempted to split the new function up in a `new` for prod and `new_with_handle` for the tests, the test can then do the waiting on the handle but prod doesn't need to have all this extra code complexity.

File: relay-server/src/utils/garbage.rs (hunk 5 above)
Comment (flub) [reply relay-server/src/utils/garbage.rs]: question is if you need this at all.  the risk you are trying to mitigate is a disposal thread that doesn't keep up and the channel growing indefinitely.  Slowing down the channel is not helpful with that, and neither is having a metric: it will only allow you to figure out eventually that there's a metric growing together with the memory.

The way to really mitigate is to make a bounded channel: if the channel is full you can log the error to sentry and you start dropping in the sender thread again.

Comment (jjbayer) [reply relay-server/src/utils/garbage.rs]: Sorry I somehow missed this comment and added the metric anyway. The bounded channel idea makes sense to me, let's see what values the metric shows and then follow up if necessary.

**Review by flub — APPROVED** [review]
It might be worth adding the bounded queue as a follow up task to do once you have some metrics on the size of it.

File: CHANGELOG.md [hunk 6]
```diff
@@ -9,6 +9,7 @@
 **Internal**:
 
 - Refactor tokio-based Addr from healthcheck to be generic. ([#1405](https://github.com/relay/pull/1405))
+- Defer dropping of projects to a background thread to speed up cache eviction. ([#1410](https://github.com/getsentry/relay/pull/1410))
```
Comment (iker-barriocanal) [inline CHANGELOG.md]: Is it worth it to explicitly mention it's for projects? Your call.
```suggestion
- Defer dropping of projects to a background thread to speed up project cache eviction. ([#1410](https://github.com/getsentry/relay/pull/1410))
```

File: relay-server/src/actors/project_cache.rs [hunk 7]
```diff
@@ -85,8 +86,18 @@ impl ProjectCache {
...
+            .projects
+            .drain_filter(|_, entry| entry.last_updated_at() + delta <= eviction_start);
+
+        // Defer dropping the projects to a dedicated thread:
+        for (_, project) in expired {
+            self.garbage_disposal.dispose(project);
+        }
+
+        // Log garbage queue size:
+        let queue_size = self.garbage_disposal.queue_size() as f64;
```
Comment (flub) [inline relay-server/src/actors/project_cache.rs:99]: this alone was enough reason for me to report queue size as u64 rather than usize :)

File: relay-server/src/statsd.rs [hunk 8]
```diff
@@ -5,12 +5,16 @@ pub enum RelayGauges {
...
+
+    /// The number of items currently in the garbage disposal queue.
+    ProjectCacheGarbageQueueSize,
 }
 
 impl GaugeMetric for RelayGauges {
     fn name(&self) -> &'static str {
         match self {
             RelayGauges::NetworkOutage => "upstream.network_outage",
+            RelayGauges::ProjectCacheGarbageQueueSize => "project_cache.garbage.queue_size",
```
Comment (flub) [inline relay-server/src/statsd.rs:17]: if all this was to get rid of the nonsensical tag, I'd have rather gone with passing a string for the tag into the constructor.  but whatever, either will do.

**Review by flub — APPROVED** [review]
what happened about the custom sampling of this metric?  did you write down the motivation to get rid of it somewhere?

**Review by jan-auer — APPROVED** [review]

File: relay-server/src/actors/project_cache.rs (hunk 7 above)
Comment (jan-auer) [reply relay-server/src/actors/project_cache.rs:99]: Can you clarify? You could equally have the queue_size as `usize` here and then cast this to an f64.

Comment (flub) [reply relay-server/src/actors/project_cache.rs:99]: Wait, why is this `f64` and not `u64`?  Only just noticed that and now I'm confused.

- Metrics are reported as `u64`.
- We don't really know what sizes a channel can get up to as this is not described AFAIK, but we can probably assume we'd crash before we reach the max size.
- That means deciding to expose it as a `usize` is arbitrary as exposing it as a `u64`.
- But now here you are lead to believe that the true value is represented by a `usize` and you're supposedly casting here.  This is misleading.

File: relay-server/src/statsd.rs (hunk 8 above)
Comment (jjbayer) [reply relay-server/src/statsd.rs:17]: The main motivation was to get rid of the `i % 100` condition for logging. The utility class now does not concern itself with statsd, but simply offers a method to query its queue size, and it's up to the caller when and how to log that information, which I think is a better separation of concerns.

**Review by iker-barriocanal — APPROVED** [review]

File: relay-server/src/statsd.rs (hunk 8 above)
Comment (flub) [reply relay-server/src/statsd.rs:17]: (note i approved so i'm not asking for changes)
The metrics are either needed or not, it shouldn't be up to the caller to think of hooking up the metrics, we're not building a library but an application (and even then, but that's a larger topic).
